### PR TITLE
Merge upstream into preveil branch

### PR DESCRIPTION
**Pull request by ianyh**

### CODE REVIEW

File: buffer.go [hunk 1]
```diff
@@ -235,6 +236,10 @@ func (c *Cache) Start(es *EventStream) {
 }
 
 func (c *Cache) Stop() {
+	if c.done == nil {
```
Comment (hu13) [inline buffer.go:239]: do you know when/how this can happen?

File: fsevents.go [hunk 2]
```diff
@@ -212,16 +197,25 @@ func (es *EventStream) Start() {
 	cbInfo := registry.Add(es)
 	es.registryID = cbInfo
 	es.uuid = GetDeviceUUID(es.Device)
-	es.start(es.Paths, cbInfo)
-	es.renameCache.Start(es)
+	err := es.start(es.Paths, cbInfo)
+	if err != nil {
+		// Remove eventstream from the registry
+		registry.Delete(es.registryID)
+		es.registryID = 0
```
Comment (hu13) [inline fsevents.go:204]: this looks right, but should we also try to clean up resources of `es.renameCache` if any? meaning should we make es.renameCache.Stop() idempotent and try to call it here?
basically, i want to make sure we can restart `EventStream` cleanly after it hits some error.

Comment (hu13) [reply fsevents.go:204]: would be good to have a test case for a restart if possible.

File: buffer.go (hunk 1 above)
Comment (ianyh) [reply buffer.go:239]: Not exactly, but I think it's probably a race in the tests against the finalizer.

**Review by hu13 — APPROVED** [review]

**Pull request merged**